src/iface/ethernetbuilder.rs

```diff
@@ -0,0 +1,154 @@
+#![deny(missing_docs)]
+
+use managed::ManagedSlice;
+
+use phy::{Device, DeviceCapabilities};
+use wire::{EthernetAddress, IpCidr};
+use wire::Ipv4Address;
+use super::{EthernetInterface, NeighborCache};
+use super::ethernet::InterfaceInner as EthernetInterfaceInner;
+
+/// A builder structure used for creating a Ethernet network
+/// interface.
+pub struct InterfaceBuilder <'b, 'c, DeviceT: for<'d> Device<'d>> {
+    device:              DeviceT,
+    device_capabilities: DeviceCapabilities,
+    ethernet_addr:       Option<EthernetAddress>,
+    neighbor_cache:      Option<NeighborCache<'b>>,
+    ip_addrs:            Option<ManagedSlice<'c, IpCidr>>,
+    ipv4_gateway:        Option<Ipv4Address>,
+}
+
+impl<'b, 'c, DeviceT> InterfaceBuilder<'b, 'c, DeviceT>
+        where DeviceT: for<'d> Device<'d> {
+    /// Create a builder used for creating a network interface using the
+    /// given device and address.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::collections::BTreeMap;
+    /// use smoltcp::iface::{EthernetInterfaceBuilder, NeighborCache};
+    /// # use smoltcp::phy::Loopback;
+    /// use smoltcp::wire::{EthernetAddress, IpCidr, IpAddress};
+    ///
+    /// let device = // ...
+    /// # Loopback::new();
+    /// let hw_addr = // ...
+    /// # EthernetAddress::default();
+    /// let neighbor_cache = // ...
+    /// # NeighborCache::new(BTreeMap::new());
+    /// let ip_addrs = // ...
+    /// # [];
+    /// let iface = EthernetInterfaceBuilder::new(device)
+    ///         .ethernet_addr(hw_addr)
+    ///         .neighbor_cache(neighbor_cache)
+    ///         .ip_addrs(ip_addrs)
+    ///         .finalize();
+    /// assert!(iface.is_some());
+    /// ```
+    pub fn new(device: DeviceT) -> InterfaceBuilder<'b, 'c, DeviceT> {
+        let caps = device.capabilities();
+        InterfaceBuilder {
+            device:              device,
+            device_capabilities: caps,
+            ethernet_addr:       None,
+            neighbor_cache:      None,
+            ip_addrs:            None,
+            ipv4_gateway:        None
+        }
+    }
+
+    /// Set the Ethernet address the interface will use.
+    ///
+    /// # Panics
```

Comment:
All these functions should probably reference the ones on `EthernetInterface` so that the docs don't go out of sync.